utils/src/main/java/software/amazon/awssdk/utils/async/SimplePublisher.java

```diff
@@ -0,0 +1,497 @@
+            entryTypesToFail.addAll(asList(QueueEntry.Type.values()));
+            subscriber.onError(failure);
+
+            while (true) {
+                QueueEntry<T> entry = eventQueue.poll();
+                if (entry == null) {
+                    break;
+                }
+                entry.resultFuture.completeExceptionally(failure);
+            }
+        } catch (Throwable t) {
+            t.addSuppressed(cause);
+            log.error(() -> "Failed while processing a failure. This could result in stuck futures.", t);
+        }
+    }
+
+    /**
+     * Ensure that {@link #rejectException} is null. If it is not, throw the exception.
+     */
+    private void checkRejectException() {
+        if (rejectException.get() != null) {
+            throw rejectException.get().get();
+        }
+    }
+
+    /**
+     * Set the {@link #rejectException}, if it is null. If it is not, throw the exception.
+     */
+    private void setRejectExceptionOrThrow(Supplier<RuntimeException> rejectedException) {
+        if (!rejectException.compareAndSet(null, rejectedException)) {
+            throw rejectException.get().get();
+        }
+    }
+
+    /**
+     * The subscription passed to the first {@link #subscriber} that subscribes to this publisher. This allows the downstream
+     * subscriber to request for more {@code onNext} calls or to {@code cancel} the stream of messages.
+     */
+    private class SubscriptionImpl implements Subscription {
+        @Override
+        public void request(long n) {
+            log.trace(() -> "Received request() with " + n);
+            if (n <= 0) {
+                // Create exception here instead of in supplier to preserve a more-useful stack trace.
+                IllegalArgumentException failure = new IllegalArgumentException("A downstream publisher requested an invalid "
+                                                                                + "amount of data: " + n);
+                rejectException.compareAndSet(null, () -> failure);
+                eventQueue.add(new OnErrorQueueEntry<>(failure));
+                entryTypesToFail.add(ON_NEXT);
+                processEventQueue();
+            } else {
+                long newDemand = outstandingDemand.addAndGet(n);
+                log.trace(() -> "Increased demand to " + newDemand);
+                processEventQueue();
+            }
+        }
+
+        @Override
+        public void cancel() {
+            log.trace(() -> "Received cancel()");
+
+            // Create exception here instead of in supplier to preserve a more-useful stack trace.
+            IllegalStateException failure = new IllegalStateException("A downstream publisher has cancelled the subscription.");
+            rejectException.compareAndSet(null, () -> failure);
+            eventQueue.add(new CancelQueueEntry<>());
+            entryTypesToFail.addAll(asList(ON_NEXT, ON_COMPLETE, ON_ERROR));
```

Comment:
> I wonder if we should just delete those events by removing the data from the queue to fail faster in case there are a lot of events in the queue.

And leave the futures incomplete? 

> Is there a use case for handling failure for individual data event?

Not that I can think of, but there is a use-case for handling success of each individual event, and it feels unnatural (to me) to have item-level success and publisher-level failure.

> A more general question, do we need entryTypesToFail? Would a isDone flag suffice?

We could combine some things for now, like ON_COMPLETE and ON_ERROR (we never fail one without failing the other), but we need to fail ON_NEXT without failing ON_ERROR, for example.

I originally had two flags - one for failing ON_NEXT and one for failing ON_COMPLETE and ON_ERROR, but there was more code and volatile reads under that design, so I went with this one. Is it confusing?

Reply:
> And leave the futures incomplete?

If we go with this approach, we'd probably change it to void and not return a future and that's why I was curious if there was a use-case for handling the result of each event. Reading it again, seems like we do(increasing window size). I guess alternatively, we could expose callback for onNext call, which is not necessarily better than this.